src/server/game/Craft/Crafting.cpp

```diff
@@ -0,0 +1,243 @@
+                }
+            }
+        }
+
+        if (reagentCountForSlot < uint32(spellSlot->ReagentCount))
+        {
+            // If we don't have enough Reagent for this slot and it is required, craft failed
+            if (reagentSlot->ReagentType == MCR_REAGENT_TYPE_REQUIRED)
+                return SpellCastResult::SPELL_FAILED_NEED_MORE_ITEMS;
+
+            // Still not enough but the reagent is optional, just skip
+            continue;
+        }
+
+        std::vector<ModifiedCraftingCategoryEntry const*> categories = sDB2Manager.GetMCRCategoryByReagentSlot(spellSlot->ModifiedCraftingReagentSlotID);
+        bool hasOneMatchingItem = false;
+
+        for (ModifiedCraftingCategoryEntry const* category : categories)
+        {
+            std::vector<ModifiedCraftingReagentItemEntry const*> reagentItems = sDB2Manager.GetMCRReagentItemByCategory(category->ID);
+
+            for (ModifiedCraftingReagentItemEntry const* reagentItem : reagentItems)
+            {
+                std::set<ItemSparseEntry const*> requiredItems = sDB2Manager.GetItemSparseByMCRReagentItem(reagentItem->ID);
+
+                for (WorldPackets::Spells::SpellCraftingReagent const& providedReagent : providedReagentItr->second)
+                {
+                    bool hasItem = std::count_if(std::cbegin(requiredItems), std::cend(requiredItems), [providedReagent](ItemSparseEntry const* itemSparse) {
+                        return itemSparse->ID == uint32(providedReagent.ItemID);
+                    });
+
+                    if (hasItem)
+                        bonusTreeIds.push_back(reagentItem->ItemBonusTreeID);
+
+                    hasOneMatchingItem = hasOneMatchingItem || hasItem;
+                }
+            }
+        }
+
+        // If any item provided do not match, probably a cheater, return
+        if (!hasOneMatchingItem)
+            return SpellCastResult::SPELL_FAILED_NEED_MORE_ITEMS;
+    }
+
+    uint32 skillLevel = GetSkillLevelForCraft();
+    uint32 difficultyPercentMatched = round(float(skillLevel) / float(craftingDifficulty) * 100.f);
+
+    uint32 qualityTier = sDB2Manager.GetCraftingQualityTierByDifficultyPercent(_craftingData->CraftingDifficultyID, difficultyPercentMatched);
+
+    uint32 craftedItemId = GetCraftedItemIdForQuality(qualityTier);
+
+    // If we didn't found anything, we have a problem. Return and don't take reagent from player.
+    if (!craftedItemId)
+        return SpellCastResult::SPELL_FAILED_ERROR;
+
+    ItemContext context = _spell->GetSpellInfo()->HasAttribute(SPELL_ATTR0_IS_TRADESKILL) ? ItemContext::Trade_Skill : ItemContext::NONE;
+
+    std::vector<int32> bonusLists;
+    for (int32 bonusTreeID : bonusTreeIds)
+    {
+        std::vector<int32> treeBonusList = ItemBonusMgr::GetBonusListsForItem(craftedItemId, { context, {}, {}, bonusTreeID, qualityTier });
+        bonusLists.insert(std::end(bonusLists), std::begin(treeBonusList), std::end(treeBonusList));
```

Comment:
this is creating unnecessary copies by inserting directly, it's better to reserve and emplace_back or std::move for that matter